### PR TITLE
Post remove ci templates

### PR DESCRIPTION
**Pull request by jneira**
Adressing some issues after the new gha workflows in #7952 

@andreabedini i started to work on this before your comments in #7592. Let me know if you want to continue the great work done there with your own pr. No problem at all with that

---
Please include the following checklist in your PR:

* [X] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).

### CODE REVIEW

File: .github/workflows/linux.yml
Comment (jneira) [inline .github/workflows/linux.yml:123]: ugh i lost the haskell setup action of both ghcs in a force push 🤦

File: .github/workflows/linux.yml
Comment (andreabedini) [inline .github/workflows/linux.yml:14]: This is ok now but I was plannig to unify linux and macos so this would be `${{ matrix.os }}`

**Comment by andreabedini** [comment]
Thanks for pinging me @jneira. My problem with testing using old GHCs is that it is very arbitrary what we are going to test.

* Why only on Linux?
* After compiling cabal with what version of GHC? (the one that GHCup marks as recommended? 8.10.7)

There's a proverbial combinatorial explosion of possible configurations to tests. I honestly do not know what we should tests.

One suggestion I could give now is that perhaps you can consider making validate-old-ghcs a separate workflow, so we can unify the main workflows (Linux / MacOS / Windows) into something simple.

**Comment by jneira** [comment]
> Thanks for pinging me @jneira. My problem with testing using old GHCs is that it is very arbitrary what we are going to test.
> 
>     * Why only on Linux?

I guess it already was only tested in linux to save ci resources and simplify the scripts. It is not comprehensive but it worked well enough until now afaik so i only keep the existing behaviour 

>     * After compiling cabal with what version of GHC? (the one that GHCup marks as recommended? 8.10.7)
> 

Again, i am keeping the existing behaviour, All builds > 8.0 should work so i hope the chosen main ghc is somewhat arbitrary. We could try 8.10.7 and see how is going (but i prefer start with the existing one to not mix things)

> There's a proverbial combinatorial explosion of possible configurations to tests. I honestly do not know what we should tests.

Hmm the same combination as before, adding the recently removed ones < 8.0 could be a good start imo

> One suggestion I could give now is that perhaps you can consider making validate-old-ghcs a separate workflow, so we can unify the main workflows (Linux / MacOS / Windows) into something simple.

I considerated it but as the original workflow made the old-ghcs job require the main one so i am keeping that here.
You can unify workflows adding macos to the matrix of the validate job. Afaics the old-ghcs does not get in the way to do it.

File: .github/workflows/linux.yml
Comment (jneira) [reply .github/workflows/linux.yml:14]: yeah thas was my idea too

**Comment by jneira** [comment]
I could not resist to add

```
# See: https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#concurrency.
concurrency:
  group: ${{ github.head_ref }}-${{ github.workflow }}
  cancel-in-progress: true
```

to all workflows, to cancel ongoing runs on new commits of the same workflow, another way to save resources

**Comment by andreabedini** [comment]
> to all workflows, to cancel ongoing runs on new commits of the same workflow, another way to save resources

Fair enough. I would prefer to see all the issues at once that having to discover them one by one but I don't have a strong opinion here.

**Comment by jneira** [comment]
> Edit: I would not say "abuse", it's the CI's job to run tests!

yeah and it continues doing it, but only for the last commit of the pr/branch if several of the are ran concurrently :wink:

**Comment by andreabedini** [comment]
I just had a little thought, on GHC 8.8.4 we pass `--solver-benchmarks` but I belive we should not.

1) it looks like it's compiled but not executed anywhere (I cannot find it looking [here](https://github.com/haskell/cabal/runs/5261070844?check_suite_focus=true), surely I must be missing it).

2) if it's a benchmark, its result might not fail/pass, and in this case I would prefer not having it as part of the CI. We could but we are trying to save on resources.

3) given the name, I assume it's benchmarking `cabal-install-solver`, in that case we could split it into a separate workflow and only run it when paths under `cabal-install-solver/` change.

my 2c.

**Comment by phadej** [comment]
> if it's a benchmark, its result might not fail/pass, and in this case I would prefer not having it as part of the CI. We could but we are trying to save on resources.

It runs solver for one package 5 times (look what `validate.sh` does). I.e. it definitely doesn't waste resources (compared to time spend figuring out why they don't build / run).

**Comment by andreabedini** [comment]
> > if it's a benchmark, its result might not fail/pass, and in this case I would prefer not having it as part of the CI. We could but we are trying to save on resources.
> 
> It runs solver for one package 5 times (look what `validate.sh` does). I.e. it definitely doesn't waste resources (compared to time spend figuring out why they don't build / run).
> 
> Don't assume, check!

:-( I didn't assume, I did check the logs and I could not find anything related to the benchmark. 

> > only run it when paths under cabal-install-solver/ change.
> 
> Bad idea. it's like testing `cabal-install` only when code in `cabal-install` changes.

And what is the argument against that?

@phadej I admire your work. I am just trying to keep things simple here, for everybody's benefit.

**Comment by andreabedini** [comment]
> Please don't remove the run of solver-benchmarks from CI. The benchmark's unit tests only cover limited functionality, so it is useful to verify that the whole benchmark can run for a single package. The benchmark runs the cabal executable, so it is testing more than just the solver.

Nobody needs to worry, I have no intention to remove it.

**Comment by Mikolaj** [comment]
> > Bad idea. it's like testing `cabal-install` only when code in `cabal-install` changes.
> 
> And what is the argument against that?

The argument is that changes in Cabal library affect how cabal-install works, so cabal-install should be tested also after any Cabal changes. Not the other way around, but I think most of our tests are cabal-install tests.

**Comment by jneira** [comment]
I am continuing the work in my fork to not take too much ci resources here: https://github.com/jneira/cabal/pull/3

**Comment by jneira** [comment]
Ok, i think i made the old-ghcs job work (see https://github.com/jneira/cabal/runs/5285931614) restoring in practice the old job.

* I only have converted it in a matrix removing the testing for the main ghc version (8.8.4)
* It builds 8.8.4 several times, the solution could be reuse the artifact generated in the validate job for that ghc. It supposes changes in the workflow (upload artifacts) and maybe in the validate script, to make it use the artefact

Thing i tried (reverse engineering @phadej's work in the way):
* Use ubuntu-lastest with no container and installing old ghcs using setup-haskell action (https://github.com/jneira/cabal/actions/runs/1874957564): it started to install , 7.10.3, 7.4.2, 7.2.2 and 7.0.4 but no 7.8.4 and 7.6.3 shrug
* Use ubuntu-18.04 with no container and installing old ghcs using setup-haskell action (https://github.com/jneira/cabal/runs/5272663327?check_suite_focus=true). It installed all old ghcs but the validate script throwed `cabal-plan: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.29' not found (required by cabal-plan)` so i deduced it needs the `-dyn` ghc flavours
* Use ubuntu-18.04 with no container and installing old ghcs (dyn flavour) using setup-haskell action (https://github.com/jneira/cabal/runs/5272663216): it semed the `dyn` flavour is not available for ubuntu-18.04 so it needs the container using xenial (16.04)
* Using xenial based container (https://github.com/jneira/cabal/runs/5284347655): The actions/checkout has problems with container using old versions of git so it has to be done manually
* So it ends with the original setup working: https://github.com/jneira/cabal/actions/runs/1880540596

So we could use ubuntu-18.04 if the validate script would support not dyn ghc flavours. I guess it is not possible for old ghc versions.

@phadej i would appreciate if you could confirm those guesses from a linux beginner

**Comment by jneira** [comment]
> Again, i am keeping the existing behaviour, All builds > 8.0 should work so i hope the chosen main ghc is somewhat arbitrary. We could try 8.10.7 and see how is going (but i prefer start with the existing one to not mix things)

will try with ghc-8.10.7

**Comment by jneira** [comment]
> > Again, i am keeping the existing behaviour, All builds > 8.0 should work so i hope the chosen main ghc is somewhat arbitrary. We could try 8.10.7 and see how is going (but i prefer start with the existing one to not mix things)
> 
> will try with ghc-8.10.7

it failed cause it has to use ghcup and haskell-setup action needs sudo in scope to install ghc with ghcup. 
Otoh the xenial container we are using has not sudo in scope: https://github.com/jneira/cabal/runs/5287480736?check_suite_focus=true

:facepalm:

**Comment by jneira** [comment]
Ok i've restored linux and macos checks and will add old-ghc ones as soon the new jobs are picked up by github ui
So the rest of prs need #8000 to be merged honouring ci

Using post-jobs make the protection rules more maintainable, you can use them as a resume of all matrix job and allows to have one check per workflow

File: .github/workflows/linux.yml
Comment (andreabedini) [inline .github/workflows/linux.yml:119]: Can you leave a comment here to clarify that this `needs` is only to limit concurrency? Otherwise I would expect this job to depends on assets produced earlier.

File: .github/workflows/linux.yml
Comment (andreabedini) [inline .github/workflows/linux.yml:121]: We might want to add a comment or two on why we are picking this old image. Apparently both you and I had to rediscoved the reason, if we don't write it down it might get lost again. (Admittedly I am about to forget again).

File: .github/workflows/linux.yml
Comment (andreabedini) [inline .github/workflows/linux.yml:158]: We might need to change this to something else otherwise [this might happen](https://github.com/haskell/cabal/runs/5289177235?check_suite_focus=true#step:17:3)

```
Unable to reserve cache with key Linux-8.8.4-3553f69b8e2c836a6106ef937b70aba70a3dd7a5, another job may be creating this cache.
```

**Comment by andreabedini** [comment]
As mentioned above, could we incorporate the details you list in https://github.com/haskell/cabal/pull/8000#issuecomment-1047646093 as comments? so next time someone touches it we won't start from scratch.

File: .github/workflows/linux.yml
Comment (jneira) [reply .github/workflows/linux.yml:158]: hmm could it have any harmful consequence?
those jobs don't need update the cache, they are reusing the main ghc one from the previous step
ideally they should reuse the artifact generated by that previous step as noted above and don't even read it
I can add a comment about that refactoring

File: .github/workflows/linux.yml
Comment (jneira) [reply .github/workflows/linux.yml:119]: I think limit concurrency is not the intent from the original `needs`, which I somewhat replicated here
The idea would be you have to ensure the main ghc is validated testing itself before using it to validate old ghcs, to not mix possible errors
It should only depend on the Linux ghc-8
8.4 from the previous job and no the entire one but I did not get to accomplish it

File: .github/workflows/linux.yml
Comment (jneira) [reply .github/workflows/linux.yml:158]: can the job use the cache in only read mode? it would make clearer it's use until such refactoring

Comment (andreabedini) [reply .github/workflows/linux.yml:158]: You might be right in that it is not a problem. Let's break it down.

AFAIU all caches in GitHub are read-only. Once they are made and associated with a key, they are not changed.

The commit was 3553f69b8e2c836a6106ef937b70aba70a3dd7a5.
1) They [fetch](https://github.com/haskell/cabal/runs/5289177235?check_suite_focus=true#step:6:14) the most recent compatible key (`Linux-8.8.4-a45a404a339570459248a81f96d41feb5f6e6922` given we give `Linux-8.8.4-` as a prefix)
2) They do their work, possibly compiling new depdendencies
3) They try to save the new cache with key `Linux-8.8.4-3553f69b8e2c836a6106ef937b70aba70a3dd7a5` and here they fail because two jobs are saving the same key at the same time.

Now, if we assume the two jobs have updated the cache in the same way, perhaps it's fine, only one needs to be able to save it.

Comment (jneira) [reply .github/workflows/linux.yml:158]: > can the job use the cache in only read mode? it would make clearer its use until such refactoring

it seems it is not possible: https://github.com/actions/cache/issues/350

**Comment by jneira** [comment]
> As mentioned above, could we incorporate the details you list in [#8000 (comment)](https://github.com/haskell/cabal/pull/8000#issuecomment-1047646093) as comments? so next time someone touches it we won't start from scratch.

I added comments about the several caveats of the job

File: .github/workflows/linux.yml
Comment (jneira) [reply .github/workflows/linux.yml:119]: > It should only depend on the Linux ghc-8.8.4 

It seems you cant depend on a specific matrix job, the job id for all of them is the same (`validate`)

**Review by andreabedini — APPROVED** [review]
LGTM. Thanks for adding all the comments! I am a bit sad we do this on Linux only since I was going to merge the platforms but alas, it will do for now.

**Comment by jneira** [comment]
> LGTM. Thanks for adding all the comments! I am a bit sad we do this on Linux only since I was going to merge the platforms but alas, it will do for now.

thanks! @Mikolaj @fgaz this need one more approval to be merged and unblock the rest of prs

**Review by Mikolaj — APPROVED** [review]
Thank you!

**Comment by Mikolaj** [comment]
@mergifyIO rebase

File: .github/workflows/bootstrap.yml
Comment (fgaz) [inline .github/workflows/bootstrap.yml:5]: why head_ref and not ref?

Comment (jneira) [reply .github/workflows/bootstrap.yml:5]: Good point `head_ref` is only defined for pull requests and workflows are triggered for direct commits too. 
The consequence is that, for consecutive direct and merge commits in master (and 3.6, 3.4 etc), they will be cancelled but the last one.
So we will lose the status of those intermmediate commits. 
There is a small probability of a green pr breaking master cause we are not enforcing the pr is up to date with master before merging it. Also there is a even small probability of merge another pr just after the offending one and we will not know the real cause is the intemmediate and no the final one.
Will change asap

**Pull request merged**
